### PR TITLE
Switch to using pillow for png as well.

### LINKED ISSUE #15165: Switch to requiring Pillow rather than having our own png wrapper?
It looks like Pillow's png handling (which is hand-written https://github.com/python-pillow/Pillow/blob/master/src/PIL/PngImagePlugin.py, not based on libpng) is faster than matplotlib's, both for writing and for reading:
```python
from time import perf_counter
import matplotlib as mpl; mpl.use("agg")
from matplotlib import pyplot as plt
import numpy as np
from PIL import Image

fig, ax = plt.subplots()
fig.canvas.draw()
t = np.asarray(fig.canvas.renderer._renderer)
for writer in ["mpl", "pil"]:
    print(f"write with {writer}")
    start = perf_counter()
    for _ in range(100):
        mpl.image.imsave("/tmp/test.png", t,
                         pil_kwargs={"mpl": None, "pil": {}}[writer])
    print("write", perf_counter() - start)
    start = perf_counter()
    for _ in range(100):
        mpl.image.imread("/tmp/test.png")
    print("mpl read", perf_counter() - start)
    start = perf_counter()
    for _ in range(100):
        with Image.open("/tmp/test.png") as image:
            mpl.image.pil_to_array(image)
    print("pil read", perf_counter() - start)
```
yields, for me:
```
write with mpl
write 1.5242545299988706
mpl read 0.9935761439992348
pil read 0.46841289500298444
write with pil
write 1.400021340996318
mpl read 0.9913013100012904
pil read 0.4612834110012045
```

https://github.com/matplotlib/matplotlib/issues/15160 reminded me that it's not always trivial to get C-level dependencies set up before building (though that specific issue is about freetype).  We already vendor agg and qhull, and have an option to download freetype (via local_freetype/MPLLOCALFREETYPE), so the sole remaining dependency that always needs to be provided by the builder is libpng.

If we just depended (as an install_requires) on Pillow for png handling instead, we would be able to drop that dependency (together with a bunch of C code).  Pillow has wheels for all three major OSes, supports PyPy, is actively developed, and has been around for a very long time.

Thoughts on making the switch?

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

Closes #15165.

Probably needs discussion as to whether the loss of precision for reading 16-bit RGB(A) pngs is a blocking issue or not.

---

The version check in _has_pil (pillow>=3.4) was dropped as the oldest
Pillow that supports Py3.6 (the oldest Python supported by Matplotlib)
is 4.0 anyways.

Some complications arise from an unfortunate design decision that
imread() returns pngs in 0-1 float dtype but others formats in integer
dtype (via Pillow), and from Pillow having not-so-great support for
16-bit images (e.g. Pillow#3041).

This makes reading of 16-bit RGB(A) PNGs (which are exposed as 0-1
floats by imread()) slightly less accurate, because Pillow coarsens them
to 8-bit first (by truncation); hence the slight change in the
pngsuite.png baseline image.  (Note that the *renderers* only have 8-bit
precision anyways (unless using mplcairo with cairo master...), but the
discrepancy comes from differences in rounding between Pillow and
Matplotlib.)

---

I have a followup PR that makes it possible to build Matplotlib on Windows on a clean python.org python with just MSVC installed (by setting MPLLOCALFREETYPE=1).

Edit: This previously depended on #15175, but is now independent of it.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by efiring** [comment]
Looks good to me, in general terms.

**Comment by anntzer** [comment]
API change note about precision loss added.

File: lib/matplotlib/backends/backend_svg.py [hunk 1]
```diff
@@ -3,12 +3,14 @@
 import gzip
 import hashlib
 import io
+from io import BytesIO
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_svg.py]: fixed

File: lib/matplotlib/backends/backend_webagg_core.py [hunk 2]
```diff
@@ -196,16 +197,16 @@ def get_diff_image(self):
 
             # TODO: We should write a new version of write_png that
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_webagg_core.py]: comment removed

File: lib/matplotlib/backends/backend_wx.py [hunk 3]
```diff
@@ -926,23 +926,21 @@ def draw(self, drawDC=None):
     def print_bmp(self, filename, *args, **kwargs):
         return self._print_image(filename, wx.BITMAP_TYPE_BMP, *args, **kwargs)
 
-    if not _has_pil:
-        def print_jpeg(self, filename, *args, **kwargs):
-            return self._print_image(filename, wx.BITMAP_TYPE_JPEG,
-                                     *args, **kwargs)
-        print_jpg = print_jpeg
+    def print_jpeg(self, filename, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_wx.py]: no, the wx backend (not wxagg!) has always relied on wx's native handling of png/tiff/jpeg/...

File: lib/matplotlib/image.py [hunk 4]
```diff
@@ -1429,24 +1411,18 @@ def imread(fname, format=None):
...
-        try:  # Try to load the image with PIL.
-            from PIL import Image
-        except ImportError:
-            raise ValueError('Only know how to handle PNG; with Pillow '
-                             'installed, Matplotlib can handle more images')
-        with Image.open(fname) as image:
-            return pil_to_array(image)
-    from matplotlib import _png
+    img_open = (
+        PIL.PngImagePlugin.PngImageFile if ext == 'png' else PIL.Image.open)
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/image.py]: No, the point is to *force* pillow to open the image as png if `ext == "png"`, regardless of what pillow may think the image is/

File: lib/matplotlib/backends/backend_wx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/matplotlib/backends/backend_wx.py]: That's not what the removed code says.

**Comment by anntzer** [comment]
rebased

File: lib/matplotlib/backends/backend_webagg_core.py [hunk 5]
```diff
@@ -194,18 +195,16 @@ def get_diff_image(self):
...
-            # TODO: We should write a new version of write_png that
-            # handles the differencing inline
-            buff = _png.write_png(
-                output.view(dtype=np.uint8).reshape(output.shape + (4,)),
-                None, compression=6, filter=_png.PNG_FILTER_NONE)
-
+            buf = BytesIO()
+            (Image.fromarray(
+                output.view(dtype=np.uint8).reshape((*output.shape, 4))
+             .save(buf, format="png")))
```
Comment (timhoffm) [inline lib/matplotlib/backends/backend_webagg_core.py]: Just style but rather:
~~~
data = output.view(dtype=np.uint8).reshape((*output.shape, 4))
Image.fromarray(data).save(buf, format="png")
~~~

File: lib/matplotlib/backends/backend_wx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (timhoffm) [reply lib/matplotlib/backends/backend_wx.py]: If I understand this correctly, @QuLogic is right, but the API change is now documented.

File: doc/api/next_api_changes/2019-09-04-AL.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Matplotlib now uses Pillow to save and read pngs
```
Comment (timhoffm) [inline doc/api/next_api_changes/2019-09-04-AL.rst]: Please adapt to the new way of documentation API changes.

Comment (anntzer) [reply doc/api/next_api_changes/2019-09-04-AL.rst]: done

File: lib/matplotlib/backends/backend_webagg_core.py (hunk 5 above)
Comment (anntzer) [reply lib/matplotlib/backends/backend_webagg_core.py]: done

**Comment by anntzer** [comment]
self-merging per the above.

**Pull request merged**